File: frontend/types/company.ts
export type CompanyRole = "owner" | "admin" | "manager" | "employee";

export type Company = {
  id: string;
  name: string;
  logo_url: string | null;
  invoice_logo_url: string | null;
  email: string | null;
  website: string | null;
  timezone: string;
  language: string;
  bin_tax_id: string | null;
  currency: string;
  invoice_prefix: string;
  invoice_number_format: string;
  next_invoice_number: number;
  tax_percentage: string;
  footer_text: string | null;
  payment_information: string | null;
  notes: string | null;
  address: string | null;
  phone: string | null;
  created_at: string;
  updated_at: string;
};

export type CompanyUpdatePayload = Partial<{
  name: string;
  email: string | null;
  website: string | null;
  timezone: string | null;
  language: string | null;
  bin_tax_id: string | null;
  currency: string | null;
  address: string | null;
  phone: string | null;
  logo_url: string | null;
  invoice_logo_url: string | null;
  invoice_prefix: string | null;
  invoice_number_format: string | null;
  tax_percentage: string | number | null;
  footer_text: string | null;
  payment_information: string | null;
  notes: string | null;
}>;

export type CompanyUser = {
  id: string;
  company_id: string;
  email: string;
  full_name: string | null;
  role: CompanyRole;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
};

export type CompanyUsersResponse = {
  items: CompanyUser[];
};

export type CompanyInvitation = {
  id: string;
  company_id: string;
  email: string;
  full_name: string | null;
  role: CompanyRole;
  expires_at: string;
  accepted_at: string | null;
  created_at: string;
  updated_at: string;
};

export type CompanyInvitationsResponse = {
  items: CompanyInvitation[];
};

export type CompanyInvitationCreateResponse = {
  invitation: CompanyInvitation;
  invite_token: string;
};

export type CompanyInvitePayload = {
  email: string;
  full_name?: string | null;
  role: Exclude<CompanyRole, "owner">;
};

export type CompanyLogoResponse = {
  url: string;
};

File: backend/app/schemas/company.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field, ValidationInfo, field_validator

from app.core.auth import Role, validate_password_policy


class CompanyRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    logo_url: str | None = None
    invoice_logo_url: str | None = None
    email: EmailStr | None = None
    website: str | None = None
    timezone: str
    language: str
    bin_tax_id: str | None = None
    currency: str
    invoice_prefix: str
    invoice_number_format: str
    next_invoice_number: int
    tax_percentage: Decimal
    footer_text: str | None = None
    payment_information: str | None = None
    notes: str | None = None
    address: str | None = None
    phone: str | None = None
    instagram: str | None = None
    director_name: str | None = None
    welcome_message: str | None = None
    receipt_signature: str | None = None
    created_at: datetime
    updated_at: datetime


class CompanyUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=120)
    email: EmailStr | None = None
    website: str | None = None
    timezone: str | None = Field(default=None, min_length=2, max_length=64)
    language: str | None = Field(default=None, min_length=2, max_length=16)
    bin_tax_id: str | None = Field(default=None, min_length=3, max_length=32)
    currency: str | None = Field(default=None, min_length=3, max_length=3)
    address: str | None = Field(default=None, max_length=255)
    phone: str | None = Field(default=None, max_length=32)
    logo_url: str | None = Field(default=None, max_length=512)
    invoice_logo_url: str | None = Field(default=None, max_length=512)
    invoice_prefix: str | None = Field(default=None, min_length=1, max_length=16)
    invoice_number_format: str | None = Field(default=None, min_length=3, max_length=64)
    tax_percentage: Decimal | None = Field(default=None, ge=0, le=100)
    footer_text: str | None = Field(default=None, max_length=500)
    payment_information: str | None = None
    notes: str | None = None
    instagram: str | None = Field(default=None, max_length=128)
    director_name: str | None = Field(default=None, max_length=120)
    welcome_message: str | None = None
    receipt_signature: str | None = None


class CompanyLogoResponse(BaseModel):
    url: str


class CompanyUserRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    email: EmailStr
    full_name: str | None = None
    role: Role
    is_active: bool
    created_at: datetime
    updated_at: datetime
    deleted_at: datetime | None = None


class CompanyUsersResponse(BaseModel):
    items: list[CompanyUserRead]


class CompanyInviteRequest(BaseModel):
    email: EmailStr
    full_name: str | None = Field(default=None, min_length=2, max_length=120)
    role: Role

    @field_validator("role")
    @classmethod
    def prevent_owner_invites(cls, value: Role) -> Role:
        if value is Role.OWNER:
            raise ValueError("Company owners are created during registration")
        return value


class CompanyUserRoleUpdateRequest(BaseModel):
    role: Role


class CompanyInvitationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    email: EmailStr
    full_name: str | None = None
    role: Role
    expires_at: datetime
    accepted_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class CompanyInvitationCreateResponse(BaseModel):
    invitation: CompanyInvitationRead
    invite_token: str


class CompanyInvitationsResponse(BaseModel):
    items: list[CompanyInvitationRead]


class CompanyInvitationAcceptRequest(BaseModel):
    token: str = Field(min_length=32)
    full_name: str = Field(min_length=2, max_length=120)
    password: str = Field(min_length=12, max_length=128)
    confirm_password: str = Field(min_length=12, max_length=128)

    @field_validator("password")
    @classmethod
    def validate_password(cls, value: str) -> str:
        validate_password_policy(value)
        return value

    @field_validator("confirm_password")
    @classmethod
    def validate_confirm_password(cls, value: str, info: ValidationInfo) -> str:
        password = info.data.get("password")
        if password is not None and value != password:
            raise ValueError("Passwords do not match")
        return value
